### ISSUE TITLE
RPS HelloWorld Problem

### ISSUE THREAD

**Issue by daveBisMe** [architecture-building-systems/revitpythonshell#110]
Daren, thanks so much for your example of making a Revit ribbon menu.

Maybe I totally misunderstood your post (http://darenatwork.blogspot.com/2013/05/deploying-rps-scripts-with.html)

I'm not sure what I did wrong, but I can't seem to get it to work.  Here are the steps I took:

- Downloaded the 6 files from your Examples folder
- Downloaded Inno Setup Compiler
- In HelloWorld.iss I changed the following:
    - AddInFilePath := ExpandConstant('{userappdata}\Autodesk\Revit\Addins\2020\HelloWorld.addin');
    - This was in 2 locations in the file and I pointed to my Revit 2020 Addins folder
    - I also created my own GUID and pasted that in the file
    - Inno Setup had me change pf32 to commonpf32 so I did, here's what I had: DefaultDirName={commonpf32}\HelloWorld

- In the buildRibbonPanel.py file I changed the following to point to where I put the files:
    - SCRIPT_PATH = r"Z:\Revit\Revit Python Shell\helloworld.py"  # noqa
    - LARGE_IMG_PATH = r"Z:\Revit\Revit Python Shell\PythonScript32x32.png"  # noqa
    - SMALL_IMG_PATH = r"Z:\Revit\Revit Python Shell\PythonScript16x16.png"  # noqa
    - DLL_FOLDER = r"C:\Users\burwell\AppData\Roaming\RevitPythonShell\2020"
- Then in Inno Setup I pressed the Run button (also tried Build/Compile)
- It ran through the steps and I ran Setup_HelloWorld.exe which created the file HelloWorld.addin in this folder: C:\Users\xxxxxxx\AppData\Roaming\Autodesk\Revit\Addins\2020

Here's what I got when I tried to open Revit:

![Error](https://user-images.githubusercontent.com/64531180/83173449-06313880-a0e7-11ea-8fd8-c745903ef09e.jpg)

I hope I gave you enough information to ask the simple question - What did I miss?

My goal is to create an executable file to contain the Python Scripts I've created and put them on each computer in our office.

I look forward to hearing from you, and thanks again for all of your valuable resources!  FYI, out the door for 2 weeks in California so I may be slow to respond.

**Comment by daren-thomas** [comment]
@daveBisMe - hmm... it's been a while since I've done this myself... let's see... (I'm following your steps here)

- so you first clicked "Deploy RPS Addin" in Revit 2020 and then selected "HelloWorld.xml", right?
- then updated "HelloWorld.iss" (changed the two "2013" to "2020")
  - NOTE: I'm not changing the `{pf}` as my InnoSetup doesn't seem to care...
  - NOTE: didn't need to change the `buildRibbonPanel.py` as it's not part of the sample
- installed the Setup_HelloWorld.exe and restarted Revit 2020
- BOOM! the exact same error message. I'm sorry. It really does seem like there's a bug here somewhere...

**Comment by daren-thomas** [comment]
@daveBisMe - yes, I had an error in the code - mainly, because I was stupid and didn't test what happens when the paths to the icon files mentioned in the RpsAddin xml file don't exist. I've fixed this with the latest push to master - it should be in the next release. If you'd rather not wait, I've attached a replacement of the RpsRuntime.dll file that has the fix - just deploy it with your `HelloWorld.dll` file.

[RpsRuntime.zip](https://github.com/architecture-building-systems/revitpythonshell/files/4746159/RpsRuntime.zip)

**Comment by daveBisMe** [comment]
Daren, I’m just getting back from vacation.  I’ll check out the RpsRuntime when I get unburied.  Thanks a ton for your help.

 

Dave

**Comment by daveBisMe** [comment]
Daren, worked great… One question however – is it easy to make a separate tab for scripts instead of using the addins tab?  I’ve got a handful of python scripts written and basically want to create my own tab.  I’ve been through the Revit API docs and it all used C# (which I know very little of).  I’ve also looked at pyRevit, which is awesome, but want to avoid loading too many scripts on end users machines. Just wondering if there was an easy way around it.

 

Thanks again.

**Comment by daren-thomas** [comment]
@daveBisMe - you have full access to the Revit API in the startup script. This article should help you figure it out: https://daren-thomas.gitbooks.io/scripting-autodesk-revit-with-revitpythonshell/content/more_control_over_the_ribbonpanel.html

**Comment by daveBisMe** [comment]
Daren, thanks once more for your help.  Looks pretty simple but what am I missing?  I’ve tried every form of syntax for my “EXAMPLES_PATH” and nothing works.  I’m sure it’s something stupid, but I’ll keep trying.

 

Here’s what I get – if you have any suggestions.  I feel bad asking, this is usually pretty simple for me to figure out.  Must still be on vacation brain or something.

 

 

'''

simple_ribbon.py - creates a ribbon panel with a single push button.

 

NOTE:

    - this MUST be set as a startup script for it to work

    - the RPS variable "EXAMPLES_PATH" must be set and contain "helloworld.py"

'''

 

# script that is run when Revit starts in the IExternalApplication.Startup event.

try:

    import os

    from RevitPythonShell.RpsRuntime import ExternalCommandAssemblyBuilder

    from Autodesk.Revit.UI import *

 

    SCRIPT_PATH = os.path.join(__vars__['S:\Users\DTB\RevitMenu\'], "helloworld.py")

    DLL_PATH = os.path.expandvars(r"%APPDATA%\Roaming\Autodesk\Revit\Addins\2020\simple_ribbon.dll")

    print 'storing external command assembly here:', DLL_PATH

 

    builder = ExternalCommandAssemblyBuilder()

    builder.BuildExternalCommandAssembly(

        DLL_PATH,

        {'HelloWorld': SCRIPT_PATH})

 

    panel = __uiControlledApplication__.CreateRibbonPanel('simple_ribbon')

 

    pbd = PushButtonData('pb_HelloWorld', 'hello, world!', DLL_PATH, 'HelloWorld')

    panel.AddItem(pbd)

 

    #__window__.Close()  # closes the window

except:

    import traceback       # note: add a python27 library to your search path first!

    traceback.print_exc()  # helps you debug when things go wrong

**Comment by daveBisMe** [comment]
Daren, ignore my last stupidity!!!  I now see what you mean by EXAMPLES_PATH.

 

Signed,

 

DumbAss…

**Comment by daren-thomas** [comment]
@daveBisMe: great that you figured it out :) notice also, that when you have a string literal with a backslash (`\`) in it, you should use the `r"this\string\has\backslashes"` version. notice how I prepended an _r_ to the string literal? that tells python not to use the backslash for escaping.

**Comment by daveBisMe** [comment]
Daren, first off – forgive me for all of the questions, but I keep running into roadblocks.  I’m going to lay out exactly what I’ve done.

 

FIRST: using Deploy RpsAddin

 

I’ve modified the HelloWorld.xml file to add SplitButtons per your documentation here: https://daren-thomas.gitbooks.io/scripting-autodesk-revit-with-revitpythonshell/content/deploying_rpsaddins/creating_the_rpsaddin_manifest.html

 

This is my new HelloWorld.xml file:

 

<?xml version="1.0" encoding="utf-8" ?>

<RpsAddin>

  <RibbonPanel text="Showing off my SplitButtons">

    <SplitButton text="SplitButton #1">

      

      <PushButton text="Hello World!" src="Z:\Revit\Revit Python Shell\PushButton1.py"/>

      <PushButton text="Goodbye cruel World!" src="Z:\Revit\Revit Python Shell\PushButton2.py"/>

    </SplitButton>

    <SplitButton text="SplitButton #2">

      <PushButton text="Gotta go" src="Z:\Revit\Revit Python Shell\PushButton3.py"/>

      <PushButton text="real fast!" src="Z:\Revit\Revit Python Shell\PushButton4.py"/>

    </SplitButton>

    <PushButton text="Enough with the Kindergarten jokes already!" src="Z:\Revit\Revit Python Shell\PushButton0.py"/>

  </RibbonPanel>

</RpsAddin>

 

I made multiple Python files (as you can see) and assigned them to each button.

 

Here’s what happens when I try to Deploy RpsAddin

 

 

From the error I gathered that something was wrong with naming each “PushButton” the same, so I did this and it deployed fine:

 

<?xml version="1.0" encoding="utf-8" ?>

<RpsAddin>

  <RibbonPanel text="Showing off my SplitButtons">

    <SplitButton text="SplitButton #1">

      

      <PushButton1 text="Hello World!" src="Z:\Revit\Revit Python Shell\PushButton2.py"/>

      <PushButton2 text="Goodbye cruel World!" src="Z:\Revit\Revit Python Shell\PushButton2.py"/>

    </SplitButton>

    <SplitButton text="SplitButton #2">

      <PushButton3 text="Gotta go" src="Z:\Revit\Revit Python Shell\PushButton3.py"/>

      <PushButton4 text="real fast!" src="Z:\Revit\Revit Python Shell\PushButton4.py"/>

    </SplitButton>

    <PushButton text="Enough with the Kindergarten jokes already!" src="Z:\Revit\Revit Python Shell\PushButton0.py"/>

  </RibbonPanel>

</RpsAddin>

 

Then I compiled the file using Inno Setup Compiler, and ran the EXE file to install the addin.  I also made sure I had your latest RpsRuntime.DLL

 

This is what I got in Revit:

 

 

The Single button worked fine but the SplitButtons were just greyed out and did nothing.

 

Another question here:  Where did the Icon come from for the single button and how do you add them to the other buttons?  I don’t see any reference anywhere to add them. I would like to add my own icons if possible.

 

SECOND:  Using Startup Script

 

Below is my Startup Script, but here are the questions I have:

 

How do you access any other python command other than HelloWorld?  I’ve tried numerous things but have no idea.  I made a dummy python file called Booger and tried it numerous ways, with and without paths in the name.  I just get an error.  The only one that seems to work is HelloWorld. Apparently I need to add a Class somewhere, but how?  See screen shot below.

 

 

"""

    - this MUST be set as a startup script for it to work

    - the RPS variable "EXAMPLES_PATH" must be set and contain "helloworld.py"

"""

import os

import clr

clr.AddReference('PresentationCore')

from System.Windows.Media.Imaging import BitmapImage

from System import Uri

from RevitPythonShell.RpsRuntime import ExternalCommandAssemblyBuilder

from Autodesk.Revit.UI import *

 

SCRIPT_PATH = os.path.join(__vars__['EXAMPLES_PATH'], "helloworld.py")

LARGE_IMG_PATH = os.path.join(__vars__['EXAMPLES_PATH'], "PythonScript32x32.png")

SMALL_IMG_PATH = os.path.join(__vars__['EXAMPLES_PATH'], "PythonScript16x16.png")

EXAMPLES_PATH = __vars__['EXAMPLES_PATH']

DLL_PATH = os.path.expandvars(r"%APPDATA%/RevitPythonShell/2020/simple_ribbon.dll")

print 'storing external command assembly here:', DLL_PATH

 

def create_ribbon_panel():

    panel = __uiControlledApplication__.CreateRibbonPanel("New Ribbon Panel")

    add_push_button(panel)

    panel.AddSeparator()

    add_split_button(panel)

    panel.AddSeparator()

 

def add_push_button(panel):

    """add push button"""

    push_button = panel.AddItem(

        PushButtonData("pb_HelloWorld", "Hello, world!",

                       DLL_PATH, "HelloWorld"))

    push_button.ToolTip = "Say hello world"

    #context_help = ContextualHelp(ContextualHelpType.Url, "http://www.autodesk.com")

    #push_button.SetContextualHelp(context_help)

 

    push_button.LargeImage = BitmapImage(Uri(LARGE_IMG_PATH))

 

def add_split_button(panel):

    """add a split button"""

    button_one = PushButtonData("pbButtonOne", "Option one",

                                DLL_PATH, "Booger")

    button_one.LargeImage = BitmapImage(Uri(os.path.join(

        EXAMPLES_PATH, 'StartupScripts', 'one_small.png')))

 

    button_two = PushButtonData("pbButtonTwo", "Option two",

                                DLL_PATH, "HelloWorld")

    button_two.LargeImage = BitmapImage(Uri(os.path.join(

        EXAMPLES_PATH, 'StartupScripts', 'two.png')))

 

    button_three = PushButtonData("pbButtonThree", "Option three",

                                DLL_PATH, "HelloWorld")

    button_three.LargeImage = BitmapImage(Uri(os.path.join(

        EXAMPLES_PATH, 'StartupScripts', 'three.png')))

 

    button_four = PushButtonData("pbButtonFour", "Option four",

                                DLL_PATH, "Booger")

    button_four.LargeImage = BitmapImage(Uri(os.path.join(

        EXAMPLES_PATH, 'StartupScripts', 'one_small.png')))

 

    split_button = panel.AddItem(SplitButtonData("splitButton", "Split"))

    split_button.AddPushButton(button_one)

    split_button.AddPushButton(button_two)

    split_button.AddPushButton(button_three)

    split_button.AddPushButton(button_four)

 

if __name__ == '__main__':

    try:

        builder = ExternalCommandAssemblyBuilder()

        builder.BuildExternalCommandAssembly(

            DLL_PATH,

            {'HelloWorld': SCRIPT_PATH})

 

        create_ribbon_panel()

    except:

        import traceback

        traceback.print_exc()

 

 

And FINALLY, one last question:  Does the ribbon panel always need to go under AddIns or can it be its own Tab?  I’m sure that would open a whole new can of worms…

 

Again, sorry for all the questions.  I’m trying my best to fully understand what you are doing but you’re on a different level than I am!  I’m not worthy!!!

 

All the best Daren

**Comment by daveBisMe** [comment]
Okay Daren, once again ignore my previous email.  I scoured the Internet for as many answers as I could find (for a second time) so I don’t bother you.

 

I got the menu’s to work using Deploy RpsAddin. At times it didn’t like the XML file even though nothing changed – weird.  I did find that the RpsRuntime you sent me would sometimes get set back to the old one when I ran InnoSetup and that would cause problems.  Not sure how that happens.  Is it copying that file from another location I’m not aware of?

 

My one question on this is how to change the icons?  I found a post saying to add “largeImage” to the XML file.  Maybe my code is wrong so here is my XML (I’ve tried both LargeImage and largeImage):

 

<?xml version="1.0" encoding="utf-8" ?>

<RpsAddin>

  <RibbonPanel text="My Ribbon">

    <SplitButton text="SplitButton #1">

      

      <PushButton text="Hello World!" src="Z:\Revit\Revit Python Shell\PushButton1.py" largeImage="Z:\Revit\Revit Python Shell\VAIcon32x32.png"/>

      <PushButton text="Goodbye cruel World!" src="Z:\Revit\Revit Python Shell\PushButton2.py"/>

    </SplitButton>

    <SplitButton text="SplitButton #2">

      <PushButton text="Gotta go" src="Z:\Revit\Revit Python Shell\PushButton3.py"/>

      <PushButton text="real fast!" src="Z:\Revit\Revit Python Shell\PushButton4.py"/>

    </SplitButton>

    <PushButton text="try this" src="Z:\Revit\Revit Python Shell\PushButton0.py"/>

  </RibbonPanel>

</RpsAddin>

 

Another question: Is Deploy RpsAddin including the reference python files in the addin or is it just reading the file from the Source location?  The reason I ask this is once I install the addin on a user’s machine – I may want to just update the python file without reloading everything.  Is that possible?

 

Now a bit about the Startup Script.

 

If using the Startup Script method, then I need to install RevitPythonShell on each computer and point to the startup script – correct?

 

Last question, then I’ll buy you a beer or two next time you’re in town!  Can you make a menu that is in its own tab using RPS, or do they always have to fall under Addins?

 

Thanks again, your help has been greatly appreciated!

**Comment by daren-thomas** [comment]
@daveBisMe - I tried formatting your previous message a bit (if you use tripple backticks, you can have a code block - makes it _much_ easier for me to read your code snippets).

As to your questions:

- if I remember correctly, the DeployRpsAddin functionality includes the scripts _into_ the DLL it creates. So no, you can't easily replace them on the user's computer. That said: Replacing a DLL is not harder than replacing a python script. If you really want to replace a python script, then have your (embedded) script import a script from the filesystem. I'll leave that as an exercise :)

- the rpsaddin manifest can also include a startup script tag - there should be an example of how to do that out there somewhere.

- creating a Panel that's not in the AddIns tab - if it's something that can be done with the RevitAPI then the startup script is where to do it. Check out this article: https://daren-thomas.gitbooks.io/scripting-autodesk-revit-with-revitpythonshell/content/more_control_over_the_ribbonpanel.html

**Comment by daveBisMe** [comment]
Daren, FINALLY getting this figured out.  Again, you are my hero.

 

Only a couple small things I’ve yet to find an answer to.

 

On my Deployed RpsAddin I am still unable to change the menu icons – I’ve tried everything.

 

Here’s a line from my XML file (forgive the formatting, I’m responding by email!)

 

    <PushButton text="Script 1" src="Z:\Revit\PushButton1.py" largeImage="Z:\Revit\VAIcon32x32.png" smallImage="Z:\Revit\VAIcon16x16.png"/>

 

Here’s what I get, no actual defined icons:

NOTE: I peeked into the DLL file that is created and didn’t see any references to the images. 

 

Also, one question on the Startup Script that creates the DLL when Revit starts (or at least I think that’s what’s happening).  I changed the script to have this line of code:

 

    DLL_PATH = os.path.expandvars(r"%APPDATA%/RevitPythonShell/2020/VintageArchonics.dll")

    …

    print 'storing external command assembly here:', DLL_PATH

 

Which I assumed would name my DLL “VintageArchonics.dll” but it doesn’t, it’s still called “simple_ribbon.dll”.  Is that something hardcoded into RPS or should I be able to name the DLL?

 

This is what I get when Revit starts up, even though I’ve changed the info:

 

    storing external command assembly here: C:\Users\myname\AppData\Roaming/RevitPythonShell/2020/simple_ribbon.dll

 

If you can help me figure out these last couple things.  I’ll never bother you again!  And, the beer count will be up to four, or 44, I lost count!

 

Have a great weekend.

**Comment by daren-thomas** [comment]
@daveBisMe I'm not sure anymore if the icons are included in the dll or if they're supposed to be shipped separately...

Regarding the dll name - this is a dll that is created at runtime since the Revit API expects external commands (the stuff that happens when you click a button in the ribbon) to be classes defined in .NET assemblies.

Setting `DLL_PATH = os.path.expandvars(r"%APPDATA%/RevitPythonShell/2020/VintageArchonics.dll")` should do the trick, _provided_ you're passing `DLL_PATH` to the `ExternalAssemblyBuilder`: 

```python
builder = ExternalCommandAssemblyBuilder()
builder.BuildExternalCommandAssembly(DLL_PATH,  {'HelloWorld': SCRIPT_PATH})
```

Are you sure you didn't change that parameter in your startup script? Also, make sure that `DLL_PATH` is not the same as the name of your RpsAddin... it's a separate file.

**Comment by daveBisMe** [comment]
Daren, BINGO!!!  We hit the jackpot.  So you know what I did for anyone else who needs it.

 

First off, the Icons WERE being copied to the Output folder when I run DeployRpsAddin.  BUT, I noticed that when running my Inno file to create and run the EXE, it would copy all files to a folder called “C:\Program Files (x86)\VA Revit Menu”.  When I looked in that folder it DID NOT contain any Icons (PNG) files, only DLL files. I manually copied the Icons to the folder and it worked.  So I guess the icons are not contained in the DLL and must be in same folder as the DLL’s.

 

I added the following line to the Inno file for each icon:

 

Source: Output_VAmenu\VAIcon32x32_1.png; DestDir: {app};

 

And then it copied each Icon to the same folder as the DLL’s.

 

So I had to have a reference in the XML file I used for DeployRpsAddin, and then point my Inno Setup file to copy the Icons from the Output folder to the correct folder for my Addin.

 

IS THAT HOW IT WAS SUPPOSED TO WORK?  I didn’t find any of this info anywhere, so I followed bad programming protocol and did Trial and Error!

 

As far as the DLL_PATH question, uh let’s just say dumbass attack.  I had copied my Startup script file to a new folder to finalize, made all my changes there, but since it was the same name as the old one, never noticed that I still had the old search path for my Startup Script in the RPS config.  One more time – dumbass!

 

So, until we chat again – all the best to you!!!

 

Dave
